flow/data_pipeline/data_pipeline.py

```diff
@@ -0,0 +1,273 @@
+    required_cols = {"time", "id", "speed", "x", "y"}
+    raw_output = raw_output.drop(set(raw_output.columns) - required_cols, axis=1)
+
+    extra_info = pd.DataFrame.from_dict(extra_info)
+    extra_info.set_index(["time", "id"])
+    raw_output = raw_output.merge(extra_info, how="left", left_on=["time", "id"], right_on=["time", "id"])
+
+    # add the partition column
+    # raw_output['partition'] = partition_name
+    raw_output = raw_output.sort_values(by=["time", "id"])
+    output_file_path = data_path[:-4]+"_trajectory.csv"
+    raw_output.to_csv(output_file_path, index=False)
+    return output_file_path
+
+
+def generate_trajectory_from_flow(data_path, extra_info, partition_name=None):
+    """Generate desired output for the trajectory_table based only on flow output.
+
+    Parameters
+    ----------
+    data_path : str
+        output file path
+    extra_info: dict
+        extra information needed in the trajectory table, collected from flow
+    partition_name: str
+        the name of the partition to put this output to
+    Returns
+    -------
+    output_file_path: str
+        the local path of the outputted csv file that should be used for
+        upload to s3 only, it does not the human readable column names and
+        will be deleted after uploading to s3. A copy of this file with all
+        the column name will remain in the ./data folder
+    """
+    extra_info = pd.DataFrame.from_dict(extra_info)
+    # extra_info["partition"] = partition_name
+    extra_info.to_csv(data_path, index=False)
+    upload_only_file_path = data_path[:-4] + "_upload" + ".csv"
+    extra_info.to_csv(upload_only_file_path, index=False, header=False)
+    return upload_only_file_path
+
+
+def upload_to_s3(bucket_name, bucket_key, file_path, only_query):
+    """Upload a file to S3 bucket.
+
+    Parameters
+    ----------
+    bucket_name : str
+        the bucket to upload to
+    bucket_key: str
+        the key within the bucket for the file
+    file_path: str
+        the path of the file to be uploaded
+    only_query: str
+        specify which query should be run on this file by lambda:
+        if empty: run none of them
+        if "all": run all available analysis query
+        if a string of list of queries: run only those mentioned in the list
+    """
+    s3 = boto3.resource("s3")
+    s3.Bucket(bucket_name).upload_file(file_path, bucket_key,
+                                       ExtraArgs={"Metadata": {"run-query": only_query}})
+    return
+
+
+def extra_init():
+    """Return the dictionary with all the feild pre-populated with empty list."""
+    extra_info = {"time_step": [], "id": [], "x": [], "y": [], "speed": [], "headway": [], "acceleration": [],
+                  "accel_without_noise": [], "realilzed_accel": [], "leader_id": [], "follower_id": [],
+                  "leader_rel_speed": [], "road_grade": [], "source_id": []}
+    return extra_info
```

Comment:
i think, since each of these fields is always populated, you can use the defaultdict class from collections